src/Distortion.cpp

```diff
@@ -162,93 +83,144 @@ std::tuple<double, double> distortionFunction(double ux, double uy,
+
+        fx = dx - fx;
+        fy = dy - fy;
+
+        distortionJacobian(x, y, jacobian, opticalDistCoeffs);
+
+        // Jxx * Jyy - Jxy * Jyx
+        double determinant = jacobian[0] * jacobian[3] - jacobian[1] * jacobian[2];
+        if (fabs(determinant) < 1E-6) {
+          ux = x;
+          uy = y;
+          //
+          // Near-zero determinant. Add error handling here.
+          //
+          //-- Just break out and return with no convergence
+          return;
+        }
+
+        x = x + (jacobian[3] * fx - jacobian[1] * fy) / determinant;
+        y = y + (jacobian[0] * fy - jacobian[2] * fx) / determinant;
       }
 
-      r_prev = r;
-      r2_prev = r * r;
-
-      // Compute new fractional distortion:
-      drOverR = opticalDistCoef[0]
-             + (r2_prev * (opticalDistCoef[1] + (r2_prev * opticalDistCoef[2])));
+      if ((fabs(fx) + fabs(fy)) <= tol) {
+        // The method converged to a root.
+        ux = x;
+        uy = y;
 
-      // Compute new estimate of r
-      r = rp + (drOverR * r_prev);
-      iteration++;
+        return;
+      }
+      // Otherwise method did not converge to a root within the maximum
+      // number of iterations
+    }
+    break;
+    // Compute undistorted focal plane coordinate given a distorted
+    // focal plane coordinate. This case works by iteratively adding distortion
+    // until the new distorted point, r, undistorts to within a tolerance of the
+    // original point, rp.
+    case INVERSE_RADIAL: {
+      const double tol = 1.0E-6;
+
+      double rp2 = (dx * dx) + (dy * dy);
+
+      if (rp2 > tol) {
+        double rp = sqrt(rp2);
+        // Compute first fractional distortion using rp
+        double drOverR = opticalDistCoeffs[0]
+                      + (rp2 * (opticalDistCoeffs[1] + (rp2 * opticalDistCoeffs[2])));
+        // Compute first distorted point estimate, r
+        double r = rp + (drOverR * rp);
+        double r_prev, r2_prev;
+        int iteration = 0;
+        do {
+          // Don't get in an end-less loop.  This algorithm should
+          // converge quickly.  If not then we are probably way outside
+          // of the focal plane.  Just set the distorted position to the
+          // undistorted position. Also, make sure the focal plane is less
+          // than 1km, it is unreasonable for it to grow larger than that.
+          if (iteration >= 15 || r > 1E9) {
+            drOverR = 0.0;
+            break;
+          }
+
+          r_prev = r;
+          r2_prev = r * r;
+
+          // Compute new fractional distortion:
+          drOverR = opticalDistCoeffs[0]
+                 + (r2_prev * (opticalDistCoeffs[1] + (r2_prev * opticalDistCoeffs[2])));
+
+          // Compute new estimate of r
+          r = rp + (drOverR * r_prev);
+          iteration++;
+        }
+        while (fabs(r * (1 - drOverR) - rp) > desiredPrecision);
+        ux = dx / (1.0 - drOverR);
+        uy = dy / (1.0 - drOverR);
```

Comment:
See my comment about the enumeration. I don't see why both adding and removing radial distortion are in this method.